omniledger/contracts/coins.go

```diff
@@ -0,0 +1,124 @@
+//    current coin instance. The argument must be a 64-bit uint in LittleEndian
+//  - transfer will send the coins given in the argument "coins" to the
+//    instance given in the argument "destination". The "coins"-argument must
+//    be a 64-bit uint in LittleEndian. The "destination" must be a 64-bit
+//    instanceID
+//  - fetch takes "coins" out of the account and returns it as an output
+//    parameter for the next instruction to interpret.
+//  - store puts the coins given to the instance back into the account.
+// You can only delete a contractCoin instance if the account is empty.
+func ContractCoin(cdb service.CollectionView, inst service.Instruction, c []service.Coin) (sc []service.StateChange, cOut []service.Coin, err error) {
+	cOut = c
+	switch {
+	case inst.Spawn != nil:
+		// Spawn creates a new coin account as a separate instance. The subID is
+		// taken from the hash of the instruction.
+		ca := service.InstanceID{
+			DarcID: inst.InstanceID.DarcID,
+			SubID:  service.NewSubID(inst.Hash()),
+		}
+		log.Lvlf3("Spawing coin to %x", ca.Slice())
+		return []service.StateChange{
+			service.NewStateChange(service.Create, ca, ContractCoinID, make([]byte, 8)),
+		}, c, nil
+	case inst.Invoke != nil:
+		// Invoke is one of "mint", "transfer", "fetch", or "store".
+		value, _, err := cdb.GetValues(inst.InstanceID.Slice())
+		if err != nil {
+			return nil, nil, err
+		}
+		coinsCurrent := binary.LittleEndian.Uint64(value)
+		coinsBuf := inst.Invoke.Args.Search("coins")
+		if coinsBuf == nil {
+			return nil, nil, errors.New("please give coins")
+		}
+		coinsArg := binary.LittleEndian.Uint64(coinsBuf)
+		var sc []service.StateChange
+		switch inst.Invoke.Command {
+		case "mint":
+			// mint simply adds this amount of coins to the account.
+			log.Lvl2("minting", coinsArg)
+			coinsCurrent += coinsArg
+		case "transfer":
+			// transfer sends a given amount of coins to another account.
+			if coinsArg > coinsCurrent {
+				return nil, nil, errors.New("not enough coins in instance")
+			}
+			coinsCurrent -= coinsArg
+			target := inst.Invoke.Args.Search("destination")
+			v, cid, err := cdb.GetValues(target)
+			if err == nil && cid != ContractCoinID {
+				err = errors.New("destination is not a coin contract")
+			}
+			if err != nil {
+				return nil, nil, err
+			}
+			targetCoin := binary.LittleEndian.Uint64(v)
+			var w bytes.Buffer
+			binary.Write(&w, binary.LittleEndian, targetCoin+coinsArg)
+			log.Lvlf3("transferring %d to %x", coinsArg, target)
+			sc = append(sc, service.NewStateChange(service.Update, service.NewInstanceID(target),
+				ContractCoinID, w.Bytes()))
+		case "fetch":
+			// fetch removes coins from the account and passes it on to the next
+			// instruction.
+			if coinsArg > coinsCurrent {
+				return nil, nil, errors.New("not enough coins in instance")
+			}
+			coinsCurrent -= coinsArg
+			cOut = append(cOut, service.Coin{Name: olCoin, Value: coinsArg})
```

Comment:
what happens if there is not store after fetch? does the coin get lost? is it ok for it to be lost?

Reply:
Caveat emptor?